File: src/model-groups.ts
import { ModelGroups } from './types.js';

const LEGACY_ALIASES: Record<string, string> = {
  haiku: 'fast',
  sonnet: 'balanced',
  opus: 'capable',
};

export function normalizeModelGroupName(value: string | undefined): string | undefined {
  if (!value) return undefined;
  const normalized = value.trim().toLowerCase().replace(/^slr\//, '');
  return LEGACY_ALIASES[normalized] ?? normalized;
}

export function normalizeModelGroups(value: unknown): ModelGroups {
  if (!value || typeof value !== 'object') return {};
  const source = value as Record<string, unknown>;
  const result: ModelGroups = {};
  for (const [key, val] of Object.entries(source)) {
    if (Array.isArray(val)) {
      result[key] = val.filter((x): x is string => typeof x === 'string');
    }
  }
  return result;
}

export function allGroupModelIds(modelGroups: ModelGroups): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const ids of Object.values(modelGroups)) {
    for (const id of ids) {
      if (!seen.has(id)) {
        seen.add(id);
        result.push(id);
      }
    }
  }
  return result;
}

export function selectedGroupModelIds(modelGroups: ModelGroups, requestedModel?: string): string[] | undefined {
  const group = normalizeModelGroupName(requestedModel);
  if (!group) return undefined;
  const ids = modelGroups[group];
  if (!ids || ids.length === 0) return undefined;
  return ids.length > 0 ? ids : undefined;
}

export function resolveDefaultGroup(modelGroups: ModelGroups, defaultGroup?: string): string | undefined {
  if (defaultGroup && modelGroups[defaultGroup]) return defaultGroup;
  const keys = Object.keys(modelGroups);
  return keys.length > 0 ? keys[0] : undefined;
}

File: model_groups.go
package sleepyrouter

import (
	"sort"
	"strings"
)

var legacyAliases = map[string]string{
	"haiku":  "fast",
	"sonnet": "balanced",
	"opus":   "capable",
}

func NormalizeModelGroupName(value string) string {
	if value == "" {
		return ""
	}
	normalized := strings.TrimPrefix(strings.ToLower(strings.TrimSpace(value)), "slr/")
	if alias, ok := legacyAliases[normalized]; ok {
		return alias
	}
	return normalized
}

func NormalizeModelGroups(value any) ModelGroups {
	groups, _ := NormalizeModelGroupsOrdered(value)
	return groups
}

func NormalizeModelGroupsOrdered(value any) (ModelGroups, []string) {
	groups := ModelGroups{}
	order := []string{}
	switch source := value.(type) {
	case map[string]any:
		for key, raw := range source {
			ids := stringsFromUnknownSlice(raw)
			if ids == nil {
				continue
			}
			groups[key] = ids
			order = append(order, key)
		}
	case map[string][]string:
		for key, ids := range source {
			groups[key] = append([]string(nil), ids...)
			order = append(order, key)
		}
	}
	// JSON decoding into map loses order. ReadConfig supplies its own order below;
	// direct map callers get deterministic lexical ordering instead of random routing.
	sort.Strings(order)
	return groups, order
}

func stringsFromUnknownSlice(value any) []string {
	switch values := value.(type) {
	case []any:
		out := make([]string, 0, len(values))
		for _, value := range values {
			if text, ok := value.(string); ok {
				out = append(out, text)
			}
		}
		return out
	case []string:
		return append([]string(nil), values...)
	default:
		return nil
	}
}

func completeGroupOrder(groups ModelGroups, preferred []string) []string {
	seen := make(map[string]bool)
	order := make([]string, 0, len(groups))
	for _, name := range preferred {
		if !seen[name] {
			if _, ok := groups[name]; ok {
				seen[name] = true
				order = append(order, name)
			}
		}
	}
	remaining := make([]string, 0, len(groups)-len(order))
	for name := range groups {
		if !seen[name] {
			remaining = append(remaining, name)
		}
	}
	sort.Strings(remaining)
	return append(order, remaining...)
}

func AllGroupModelIDs(groups ModelGroups, groupOrder ...string) []string {
	seen := make(map[string]bool)
	result := []string{}
	for _, group := range completeGroupOrder(groups, groupOrder) {
		for _, id := range groups[group] {
			if !seen[id] {
				seen[id] = true
				result = append(result, id)
			}
		}
	}
	return result
}

func SelectedGroupModelIDs(groups ModelGroups, requestedModel string) []string {
	group := NormalizeModelGroupName(requestedModel)
	ids, ok := groups[group]
	if group == "" || !ok || len(ids) == 0 {
		return nil
	}
	return ids
}

func ResolveDefaultGroup(groups ModelGroups, defaultGroup string, groupOrder ...string) string {
	if defaultGroup != "" {
		if _, ok := groups[defaultGroup]; ok {
			return defaultGroup
		}
	}
	order := completeGroupOrder(groups, groupOrder)
	if len(order) == 0 {
		return ""
	}
	return order[0]
}
